main/webapp/modules/core/scripts/dialogs/expression-preview-dialog.js

```diff
@@ -482,3 +524,62 @@ ExpressionPreviewDialog.Widget.prototype._renderPreview = function(expression, d
         }
     }
 };
+
+ExpressionPreviewDialog.Widget.prototype._renderClusters = function(clusters) {
+    var container = this._elmts.expressionPreviewClustersContainer.empty();
+    
+    if (clusters.length > 0) {
+        var table = $('<table></table>')
+            .addClass("clustering-dialog-preview-table")
+            .appendTo($('<div>').addClass("clusters-preview-table-wrapper").appendTo(container))[0];
+
+        var trHead = table.insertRow(table.rows.length);
+        trHead.className = "header";
+        $(trHead.insertCell(0)).text("Number");
+        $(trHead.insertCell(1)).text("Clusters");
+
+        var entryTemplate = document.createElement('a');
+
+        var renderCluster = function(cluster, index) {
+            var tr = table.insertRow();
+            tr.className = index % 2 === 0 ? "odd" : "even"; // TODO: Unused?
+
+            var ul = document.createElement('ul');
+            ul.style.listStyleType = 'none';
+            var choices = cluster.choices;
+            for (let c = 0; c < choices.length; c++) {
+                let choice = choices[c];
+                var li = document.createElement('li');
+
+                var entry = entryTemplate.cloneNode();
+                entry.textContent = choice.v.toString().replaceAll(' ', '\xa0');
+                li.append(entry);
+
+                if (choice.c > 1) { 
+                  $('<span></span>').text($.i18n("core-dialogs/cluster-rows", choice.c)).addClass("clustering-dialog-preview-count").appendTo(li);
+                }
+                ul.append(li);
+            }
+
+            $('<span>' + (index+1) +'.</span>').appendTo(tr.insertCell(0));
+
+            $(tr.insertCell(1))
+                .append(ul);
+                        
+            return choices.length;
+        };
+
+        var maxRenderRows = parseInt(
+            Refine.getPreference("ui.clustering.choices.limit", 5000)
```

Comment:
Because it's a preview it might make sense to use a lower cap (perhaps not configurable?). But it would mostly be useful if the DOM rendering is a bottleneck, which I am not sure of.

Reply:
I used the debugger to test the Cluster preview with a distance expression set to `value1.length() - value2.length()` on this [dataset](https://github.com/user-attachments/files/16568521/lieux-de-tournage-a-paris.csv). It generated over 5000 clustering choices and the clusters were displayed quickly without any slowdowns.

![](https://github.com/user-attachments/assets/16df5d42-20ed-4b38-9d73-d1ea3f91dc74)